sdks/python/apache_beam/io/gcp/bigquery_file_loads.py

```diff
@@ -0,0 +1,589 @@
+      writer.close()
+    self._destination_to_file_writer = {}
+
+
+class WriteGroupedRecordsToFile(beam.DoFn):
+  """Receives collection of dest-iterable(records), writes it to files.
+
+  This is different from ``WriteRecordsToFile`` because it receives records
+  grouped by destination. This means that it's not necessary to keep multiple
+  file descriptors open, because we know for sure when records for a single
+  destination have been written out.
+
+  Experimental; no backwards compatibility guarantees.
+  """
+
+  def __init__(self, max_file_size=_DEFAULT_MAX_FILE_SIZE,
+               coder=None):
+    self.max_file_size = max_file_size
+    self.coder = coder or bigquery_tools.RowAsDictJsonCoder()
+
+  def process(self, element, file_prefix):
+    destination = element[0]
+    rows = element[1]
+
+    writer = None
+
+    for row in rows:
+      if writer is None:
+        (file_path, writer) = _make_new_file_writer(file_prefix, destination)
+        yield (destination, file_path)
+
+      writer.write(self.coder.encode(row))
+      writer.write(b'\n')
+
+      if writer.tell() > self.max_file_size:
+        writer.close()
+        writer = None
+
+
+class TriggerCopyJobs(beam.DoFn):
+  """Launches jobs to copy from temporary tables into the main target table.
+
+  When a job needs to write to multiple destination tables, or when a single
+  destination table needs to have multiple load jobs to write to it, files are
+  loaded into temporary tables, and those tables are later copied to the
+  destination tables.
+
+  This transform emits (destination, job_reference) pairs.
+  """
+  def __init__(self,
+               create_disposition=None,
+               write_disposition=None,
+               test_client=None,
+               temporary_tables=False):
+    self.create_disposition = create_disposition
+    self.write_disposition = write_disposition
+    self.test_client = test_client
+    self.temporary_tables = temporary_tables
+
+  def start_bundle(self):
+    self.bq_wrapper = bigquery_tools.BigQueryWrapper(client=self.test_client)
+
+  def process(self, element, job_name_prefix=None):
+    destination = element[0]
+    job_reference = element[1]
+
+    if not self.temporary_tables:
+      # If we did not use temporary tables, then we do not need to trigger any
+      # copy jobs.
+      return
+
+    copy_to_reference = bigquery_tools.parse_table_reference(destination)
+    if copy_to_reference.projectId is None:
+      copy_to_reference.projectId = vp.RuntimeValueProvider.get_value(
+          'project', str, '')
+
+    copy_from_reference = bigquery_tools.parse_table_reference(destination)
+    copy_from_reference.tableId = job_reference.jobId
+    if copy_from_reference.projectId is None:
+      copy_from_reference.projectId = vp.RuntimeValueProvider.get_value(
+          'project', str, '')
+
+    copy_job_name = '%s_copy_%s_to_%s' % (
+        job_name_prefix,
+        _bq_uuid('%s:%s.%s' % (copy_from_reference.projectId,
```

Comment:
Will we end up creating the same UUID for the same destination table ? Will this result in jobs being rejected when we try to execute multiple jobs against the same destination table ? It'll be interesting to see what Java SDK currently does in this case.

Reply:
There is now an index for jobs to the same table, and this id is not regenerated.